How many late days are you using for this assignment? 

In [3]:
# Imports and pip installations (if needed)
import pandas as pd

# Part 1: Load the dataset

In [5]:
# Load the given datasets
ckd_numerical = pd.read_csv("chronic_kidney_disease_numerical.csv")
ckd_categorical = pd.read_csv("chronic_kidney_disease_categorical.csv")
# Print the data
print(ckd_numerical.head())
print(ckd_categorical.head())

   unique_id   age    bp    bgr     bu   sc    sod  pot  hemo   pcv    wbcc  \
0     203694  38.0  80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0  7300.0   
1     938027  43.0  60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0  7200.0   
2     421471  37.0  60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0  5500.0   
3     764115  70.0  90.0  144.0  125.0  4.0  136.0  4.6  12.0  37.0  8200.0   
4     240975  47.0  80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN     NaN   

   rbcc  
0   6.4  
1   5.5  
2   5.7  
3   4.5  
4   NaN  
   unique_id   al   su       rbc        pc         pcc          ba  htn   dm  \
0     224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes  yes   
1     992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   no   
2     308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   no   
3     450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   no   
4     881763  0.0  0.0    normal    normal  notpresent  notpresen

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer: The data is a collection of various health/medical markers associated with Chronic Kidney Disease. Each column represents a different vital sign/condition, and each row represents a different person, represented by a unique ID number. Some of the features measure health stats numerically, others measure it categorically. For example, age, white blood cell count (wbcc), and hemoglobin (hemo) are all measured numerically. While, red blood cell (rbc), pus cell (pc), and appetite (appet) are all measured categorically.

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [9]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
num_duplicates = ckd_numerical.duplicated().sum()
print(f"Total duplicate rows in numerical dataset: {num_duplicates}")

# Drop these duplicate rows
ckd_numerical = ckd_numerical.drop_duplicates()

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
cat_duplicates = ckd_categorical.duplicated().sum()
print(f"Total duplicate rows in categorical dataset: {cat_duplicates}")
ckd_categorical = ckd_categorical.drop_duplicates()


Total duplicate rows in numerical dataset: 0
Total duplicate rows in categorical dataset: 0


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer:

In [10]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
combined_dataset = pd.merge(ckd_numerical, ckd_categorical, on="unique_id", how="inner")

#Print the combined dataset
print(combined_dataset.head())


   unique_id   age    bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0     203694  38.0  80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0  ...   
1     938027  43.0  60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0  ...   
2     421471  37.0  60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0  ...   
3     764115  70.0  90.0  144.0  125.0  4.0  136.0  4.6  12.0  37.0  ...   
4     240975  47.0  80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN  ...   

       pc         pcc          ba  htn   dm cad appet   pe ane  Target  
0  normal  notpresent  notpresent   no   no  no  good   no  no  notckd  
1  normal  notpresent  notpresent   no   no  no  good   no  no  notckd  
2  normal  notpresent  notpresent   no   no  no  good   no  no  notckd  
3  normal  notpresent  notpresent  yes  yes  no  poor  yes  no     ckd  
4  normal  notpresent  notpresent   no   no  no  good   no  no  notckd  

[5 rows x 25 columns]


## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [11]:
# Calculate the percentage of rows that contain atleast one missing value
missing_rows = combined_dataset.isnull().any(axis=1).sum()
total_rows = len(combined_dataset)
percentage = (missing_rows / total_rows) * 100

# Print %
print(f"Percentage of rows that contain at least one missing value: {percentage:.2f}%")

# Drop these rows from the dataset
scrubbed_dataset = combined_dataset.dropna()

# Print the Dataset
print(scrubbed_dataset.head())

Percentage of rows that contain at least one missing value: 60.77%
    unique_id   age    bp    bgr    bu   sc    sod  pot  hemo   pcv  ...  \
0      203694  38.0  80.0   99.0  19.0  0.5  147.0  3.5  13.6  44.0  ...   
1      938027  43.0  60.0  108.0  25.0  1.0  144.0  5.0  17.8  43.0  ...   
2      421471  37.0  60.0  111.0  35.0  0.8  135.0  4.1  16.2  50.0  ...   
7      109053  64.0  60.0  106.0  27.0  0.7  150.0  3.3  14.4  42.0  ...   
13     349892  59.0  70.0  424.0  55.0  1.7  138.0  4.5  12.6  37.0  ...   

          pc         pcc          ba  htn   dm  cad appet  pe ane  Target  
0     normal  notpresent  notpresent   no   no   no  good  no  no  notckd  
1     normal  notpresent  notpresent   no   no   no  good  no  no  notckd  
2     normal  notpresent  notpresent   no   no   no  good  no  no  notckd  
7     normal  notpresent  notpresent   no   no   no  good  no  no  notckd  
13  abnormal  notpresent  notpresent  yes  yes  yes  good  no  no     ckd  

[5 rows x 25 column

## Part 3.4: Sort the dataset according to the Labels

In [12]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
sorted_dataset = scrubbed_dataset.sort_values(by='Target')
sorted_dataset = sorted_dataset.reset_index(drop=True)

# Print the dataset
print(sorted_dataset.head())

   unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
1     137148  73.0  100.0  295.0   90.0  5.6  140.0  2.9   9.2  30.0  ...   
2     484175  53.0   90.0   70.0  107.0  7.2  114.0  3.7   9.5  29.0  ...   
3     343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
4     532520  60.0   90.0  105.0   53.0  2.3  136.0  5.2  11.1  33.0  ...   

         pc         pcc          ba  htn   dm  cad appet   pe  ane Target  
0  abnormal     present     present   no   no   no  good   no  yes    ckd  
1  abnormal     present  notpresent  yes  yes  yes  poor   no   no    ckd  
2  abnormal     present  notpresent  yes  yes   no  poor   no  yes    ckd  
3  abnormal  notpresent  notpresent  yes  yes  yes  poor  yes  yes    ckd  
4    normal  notpresent  notpresent   no   no   no  good   no   no    ckd  

[5 rows x 25 columns]


## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [14]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
encoded_dataset = pd.get_dummies(sorted_dataset, columns=categorical_columns, drop_first=True)
encoded_dataset = encoded_dataset.astype(int)

# Print the dataset
print(encoded_dataset.head())

   unique_id  age   bp  bgr   bu  sc  sod  pot  hemo  pcv  ...  pc_normal  \
0     474407   21   90  107   40   1  125    3     8   23  ...          0   
1     137148   73  100  295   90   5  140    2     9   30  ...          0   
2     484175   53   90   70  107   7  114    3     9   29  ...          0   
3     343710   61   80  173  148   3  135    5     7   24  ...          0   
4     532520   60   90  105   53   2  136    5    11   33  ...          1   

   pcc_present  ba_present  htn_yes  dm_yes  cad_yes  appet_poor  pe_yes  \
0            1           1        0       0        0           0       0   
1            1           0        1       1        1           1       0   
2            1           0        1       1        0           1       0   
3            0           0        1       1        1           1       1   
4            0           0        0       0        0           0       0   

   ane_yes  Target_notckd  
0        1              0  
1        0              

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: The downstream modeling would not be equivalent since a single column can only effectively store truly binary data. If our condiment example can take 5 separate valeus then binary numbers of 0 and 1 wouldn't be able to represent each of the five different options in just one column/cell.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [15]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
def remove_outliers(df, numerical_columns):
    for col in numerical_columns:
        mean = df[col].mean()
        std = df[col].std()
        lower_bound = mean - (3 * std)
        upper_bound = mean + (3 * std)

        df = df[(df[col] >= lower_bound) & (df[col] <= upper_bound)]
    
    return df

cleaned_dataset = remove_outliers(encoded_dataset, numerical_columns)

# Print the dataset
print(cleaned_dataset.head())

   unique_id  age   bp  bgr   bu  sc  sod  pot  hemo  pcv  ...  pc_normal  \
0     474407   21   90  107   40   1  125    3     8   23  ...          0   
1     137148   73  100  295   90   5  140    2     9   30  ...          0   
3     343710   61   80  173  148   3  135    5     7   24  ...          0   
4     532520   60   90  105   53   2  136    5    11   33  ...          1   
6     546225   46   60  163   92   3  141    4     9   28  ...          1   

   pcc_present  ba_present  htn_yes  dm_yes  cad_yes  appet_poor  pe_yes  \
0            1           1        0       0        0           0       0   
1            1           0        1       1        1           1       0   
3            0           0        1       1        1           1       1   
4            0           0        0       0        0           0       0   
6            0           0        1       1        0           0       0   

   ane_yes  Target_notckd  
0        1              0  
1        0              

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [16]:
# Normalize the all Numerical Attributes in the dataset.
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
cleaned_dataset[numerical_columns] = scaler.fit_transform(cleaned_dataset[numerical_columns])


# Print the dataset
print(cleaned_dataset.head())

   unique_id       age    bp       bgr        bu     sc       sod       pot  \
0     474407  0.202703  0.75  0.158798  0.196078  0.125  0.166667  0.333333   
1     137148  0.905405  1.00  0.965665  0.522876  0.625  0.666667  0.000000   
3     343710  0.743243  0.50  0.442060  0.901961  0.375  0.500000  1.000000   
4     532520  0.729730  0.75  0.150215  0.281046  0.250  0.533333  1.000000   
6     546225  0.540541  0.00  0.399142  0.535948  0.375  0.700000  0.666667   

   hemo       pcv  ...  pc_normal  pcc_present  ba_present  htn_yes  dm_yes  \
0   0.1  0.000000  ...          0            1           1        0       0   
1   0.2  0.225806  ...          0            1           0        1       1   
3   0.0  0.032258  ...          0            0           0        1       1   
4   0.4  0.322581  ...          1            0           0        0       0   
6   0.2  0.161290  ...          1            0           0        1       1   

   cad_yes  appet_poor  pe_yes  ane_yes  Target_no

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: Although it may sound counterintuitive, the 'unique_id' row is unnecessary for our analysis. The goal of this dataset is to examine patterns and relationships between individuals with CKD and those without, rather than identifying specific individuals. Since the data is being used for general analysis and modeling rather than tracking individuals over time, retaining the 'unique_id' does not add value to the predictive process.

In [17]:
#Remove that column
columns_to_remove = ['unique_id']
cleaned_dataset = cleaned_dataset.drop(columns=columns_to_remove)

# Print the dataset
print(cleaned_dataset.head())

        age    bp       bgr        bu     sc       sod       pot  hemo  \
0  0.202703  0.75  0.158798  0.196078  0.125  0.166667  0.333333   0.1   
1  0.905405  1.00  0.965665  0.522876  0.625  0.666667  0.000000   0.2   
3  0.743243  0.50  0.442060  0.901961  0.375  0.500000  1.000000   0.0   
4  0.729730  0.75  0.150215  0.281046  0.250  0.533333  1.000000   0.4   
6  0.540541  0.00  0.399142  0.535948  0.375  0.700000  0.666667   0.2   

        pcv      wbcc  ...  pc_normal  pcc_present  ba_present  htn_yes  \
0  0.000000  0.653226  ...          0            1           1        0   
1  0.225806  0.217742  ...          0            1           0        1   
3  0.032258  0.395161  ...          0            0           0        1   
4  0.322581  0.500000  ...          1            0           0        0   
6  0.161290  0.830645  ...          1            0           0        1   

   dm_yes  cad_yes  appet_poor  pe_yes  ane_yes  Target_notckd  
0       0        0           0       0 

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [18]:
# Export the dataframe to a new csv file
cleaned_dataset.to_csv("cleaned_chronic_kidney_disease.csv", index=False)

# Export the dataframe to a new json file
cleaned_dataset.to_json("cleaned_chronic_kidney_disease.json", orient="records", indent=4)

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

Paste here: | Col1         | Col2 | Col3         | Col4         | Col5 | Col6         | Col7         | Col8 | Col9         | Col10        | Col11 | Col12 | Col13 | Col14 | Col15 | Col16 | Col17 | Col18 | Col19 | Col20 | Col21 | Col22 | Col23 | Col24 | Col25 |
|--------------|------|--------------|--------------|------|--------------|--------------|------|--------------|---------------|-------|-------|-------|-------|-------|-------|-------|-------|-------|-------|-------|-------|-------|-------|-------|
| 0.2027027027 | 0.75 | 0.1587982833 | 0.1960784314 | 0.125| 0.1666666667 | 0.3333333333 | 0.1  | 0.0          | 0.6532258065  | 0.0   | 4     | 0     | 1     | 0     | 1     | 1     | 0     | 0     | 0     | 0     | 0     | 1     | 0     |
| 0.9054054054 | 1.0  | 0.9656652361 | 0.522875817  | 0.625| 0.6666666667 | 0.0          | 0.2  | 0.2258064516 | 0.2177419355  | 0.0   | 3     | 2     | 0     | 0     | 1     | 0     | 1     | 1     | 1     | 1     | 0     | 0     | 0     |
| 0.7432432432 | 0.5  | 0.4420600858 | 0.9019607843 | 0.375| 0.5         | 1.0          | 0.0  | 0.0322580645 | 0.3951612903  | 0.0   | 2     | 0     | 0     | 0     | 0     | 0     | 1     | 1     | 1     | 1     | 1     | 1     | 0     |
| 0.7297297297 | 0.75 | 0.1502145923 | 0.2810457516 | 0.25 | 0.5333333333 | 1.0          | 0.4  | 0.3225806452 | 0.5          | 0.333 | 2     | 0     | 0     | 1     | 0     | 0     | 0     | 0     | 0     | 0     | 0     | 0     | 0     |
| 0.5405405405 | 0.0  | 0.3991416309 | 0.5359477124 | 0.375| 0.7         | 0.6666666667 | 0.2  | 0.1612903226 | 0.8306451613  | 0.0   | 1     | 0     | 1     | 1     | 0     | 0     | 1     | 1     | 0     | 0     | 0     | 0     | 0     |
| 0.6756756757 | 0.75 | 0.2532188841 | 0.6339869281 | 0.75 | 0.3666666667 | 0.6666666667 | 0.2  | 0.1935483871 | 0.1693548387  | 0.0   | 2     | 0     | 0     | 0     | 0     | 0     | 1     | 0     | 0     | 0     | 0     | 0     | 0     |
| 0.7162162162 | 0.5  | 1.0          | 0.1633986928 | 0.125| 0.0666666667 | 0.3333333333 | 0.3  | 0.3870967742 | 0.5322580645  | 0.333 | 1     | 0     | 0     | 1     | 0     | 0     | 0     | 1     | 0     | 1     | 0     | 0     | 0     |
| 0.7297297297 | 0.0  | 0.9356223176 | 0.1699346405 | 0.125| 0.3333333333 | 0.3333333333 | 0.0  | 0.064516129 | 0.8790322581  | 0.0   | 3     | 1     | 1     | 0     | 1     | 0     | 1     | 0     | 0     | 1     | 0     | 1     | 0     |
| 0.0          | 0.0  | 0.1030042918 | 0.3725490196 | 0.125| 0.5         | 0.6666666667 | 0.2  | 0.2258064516 | 1.0          | 0.333 | 4     | 0     | 0     | 0     | 0     | 1     | 0     | 0     | 0     | 1     | 0     | 0     | 0     |
| 0.8783783784 | 0.0  | 0.2060085837 | 0.7516339869 | 0.625| 0.5333333333 | 0.6666666667 | 0.4  | 0.3870967742 | 0.8790322581  | 0.333 | 4     | 0     | 1     | 1     | 0     | 0     | 1     | 1     | 0     | 1     | 1     | 0     | 0     |
| 0.8513513514 | 0.25 | 0.6180257511 | 0.5620915033 | 0.75 | 0.0         | 0.3333333333 | 0.2  | 0.1612903226 | 0.5806451613  | 0.0   | 4     | 3     | 1     | 0     | 1     | 1     | 1     | 1     | 1     | 0     | 1     | 1     | 0     |
| 0.8783783784 | 0.25 | 0.6394849785 | 0.4705882353 | 0.375| 0.4333333333 | 0.6666666667 | 0.3  | 0.3225806452 | 0.1048387097  | 0.0   | 3     | 0     | 1     | 0     | 1     | 1     | 1     | 1     | 1     | 0     | 0     | 0     | 0     |
| 0.7837837838 | 0.0  | 0.7253218884 | 0.3137254902 | 0.5  | 0.5666666667 | 1.0          | 0.2  | 0.1935483871 | 0.2580645161  | 0.0   | 4     | 1     | 0     | 0     | 0     | 1     | 1     | 1     | 0     | 1     | 1     | 0     | 0     |
| 0.6621621622 | 0.5  | 0.6180257511 | 0.4117647059 | 0.375| 0.5666666667 | 0.6666666667 | 0.3  | 0.3548387097 | 0.25          | 0.0   | 3     | 1     | 1     | 0     | 1     | 1     | 1     | 1     | 0     | 0     | 1     | 0     | 0     |
| 0.7702702703 | 1.0  | 0.9012875536 | 0.1633986928 | 0.375| 0.7666666667 | 0.3333333333 | 0.6  | 0.5483870968 | 0.4435483871  | 0.333 | 2     | 2     | 1     | 1     | 0     | 1     | 1     | 0     | 1     | 0     | 0     | 0     | 0     |


** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

Answer: At the end of the day, LLMs won't be flawless, and can end up making computational mistakes, or hallucinating fake responses to fill in the blanks of their argument or data. Because of this, it's imperative that whoever is using an LLM asks clear questions, scruitinizes responses returned by LLMs, and fact-checks the work produced. Hopefully as LLMs improve the necessity for these steps will diminish, but as of now these are crucial steps in order to use LLMs in a reliable way. That being said, this extra work of checking the output of LLMs allows for us, the person providing the LLM with an input, to remain involved in the process of cleaning and analzing data since it's necessary for us to know our data well in order to proof read outputs provided by LLMS.

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [ ]:
# Code to rename all the columns in the dataset
# Define the mapping of old column names to new column names
column_renames = {
    "age": "age",
    "bp": "blood pressure",
    "sg": "specific gravity",
    "al": "albumin",
    "su": "sugar",
    "rbc": "red blood cells",
    "pc": "pus cell",
    "pcc": "pus cell clumps",
    "ba": "bacteria",
    "bgr": "blood glucose random",
    "bu": "blood urea",
    "sc": "serum creatinine",
    "sod": "sodium",
    "pot": "potassium",
    "hemo": "hemoglobin",
    "pcv": "packed cell volume",
    "wc": "white blood cell count",
    "rc": "red blood cell count",
    "htn": "hypertension",
    "dm": "diabetes mellitus",
    "cad": "coronary artery disease",
    "appet": "appetite",
    "pe": "pedal edema",
    "ane": "anemia",
}

cleaned_dataset = cleaned_dataset.rename(columns=column_renames)

print(cleaned_dataset.head())

        age  blood pressure  blood glucose random  blood urea  \
0  0.202703            0.75              0.158798    0.196078   
1  0.905405            1.00              0.965665    0.522876   
3  0.743243            0.50              0.442060    0.901961   
4  0.729730            0.75              0.150215    0.281046   
6  0.540541            0.00              0.399142    0.535948   

   serum creatinine    sodium  potassium  hemoglobin  packed cell volume  \
0             0.125  0.166667   0.333333         0.1            0.000000   
1             0.625  0.666667   0.000000         0.2            0.225806   
3             0.375  0.500000   1.000000         0.0            0.032258   
4             0.250  0.533333   1.000000         0.4            0.322581   
6             0.375  0.700000   0.666667         0.2            0.161290   

       wbcc  ...  pc_normal  pcc_present  ba_present  htn_yes  dm_yes  \
0  0.653226  ...          0            1           1        0       0   
1  0.2

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [21]:
# Converted SQL to Pandas code
target_counts = cleaned_dataset.groupby("Target_notckd").size().reset_index(name="count")

print(target_counts)

   Target_notckd  count
0              0     22
1              1    111


Code Description: This code analyzes the "Target_notckd" column to count how many individuals have or do not have Chronic Kidney Disease (CKD). The results show that 22 people have a "False" or 0 value for "not having CKD", meaning they do have CKD. Meanwhile, 111 people have a "True" or 1 value for "not having CKD", indicating they do not have CKD.